### PR TITLE
Introduce $*WARNINGS and ENV<RAKU_WARNINGS>

### PR DESCRIPTION
**Pull request by lizmat**
Introduces 5 new classes that provide different behaviours for warnings

-  CX::Warn::Quietly - do not show warnings, similar to quietly
-  CX::Warn::Fatal - immediately throw a warning as an exception
-  CX::Warn::Verbose - immediately show warnings with complete backtrace
-  CX::Warn::Collect - collect warnings until END, then show with frequencies
-  CX::Warn::Debug - show warnings with complete backtrace and debugging info

The behaviour can be set by setting the new dynamic variable `$*WARNINGS` to any of these classes (CX::Warn for normal behaviour), with the exception of `CX::Warn::Debug` which *must* be instantiated.

The default behaviour of a process can be set with the `RAKU_WARNINGS` environment variable, which should have any of "quietly", "fatal", "verbose", or "collect" specified. No specification defaults to the current behaviour.

Custom behaviour can also be made by creating one's own `CX::Warn::Foo` class, which should have a WARN method taking the warning message and a `Backtrace` object. That method is then expected to do the right thing, and return `Nil`.  Such custom classes can also be activated with `RAKU_WARNINGS=foo`.

I also tried making this a v6.e feature only, but that ran into all sorts of scoping issues. So until these are better understood, this is now implemented as a 6.c feature.

The `CX::Warn::Debug` class is special in that it must be instantiated, and as such cannot be activated from an environment variable.  The `.new` method takes any number of *variables* and will show the name of the variable and its value every time a warning is issued.

    my $foo = 42;
    my $*WARNINGS = CX::Warn::Debug.new($foo);
    warn "bar";

will note:

    $foo = 42
    bar
      in ...

### CODE REVIEW

**Comment by vrurg** [comment]
I don't like the idea of introducing warnings which are not control exceptions. Though not postulated (or I just've never seen it specced in either way),  but a warning is anticipated to be a control exception. And it has to remain such, in my view. Especially considering `CX::` namespace which stands for `Control eXception` and it wouldn't be a good thing to contaminate it with non-exceptional entities. 

Thus, I'd be for different handling of warning exceptions, not for cardinal change in their behavior.

I see the point in the feature and `fatal` warnings is something I did miss on a couple of occasions. But my view would be that it is to be implemented in other way around: by delegating the final handling of a waning to the `CX::Warn` class itself. It could, say, be method `HANDLE` on the class. All other `CX::Warn::*` would then inherit from it. The method could immediately throw for `CX::Warn::Fatal`, or do nothing for `CX::Warn::Quiet`.

BTW, this may need some prior discussion in `problem-solving`.

File: src/core.c/Exception.pm6 [hunk 1]
```diff
@@ -3023,7 +3067,7 @@ my class X::Numeric::Underflow is Exception {
 
 my class X::Numeric::Uninitialized is Exception {
     has Numeric $.type;
-    method message() { "Use of uninitialized value of type " ~ $!type.^name ~ " in numeric context" }
+    method message() { nqp::say("here"); put Backtrace.new; "Use of uninitialized value of type " ~ $!type.^name ~ " in numeric context" }
```
Comment (vrurg) [inline src/core.c/Exception.pm6:3070]: Apparently, a debug leftover?

Comment (lizmat) [reply src/core.c/Exception.pm6:3070]: Oops, indeed  :-)

**Comment by lizmat** [comment]
> I don't like the idea of introducing warnings which are not control exceptions.

So why *are* warnings control exceptions?  Historically, it seems so that they can be caught indeed.  But the catching of a warning in practice is pretty complicated, needing a `CONTROL` block.

**Comment by lizmat** [comment]
Looking at the implementation of `quietly`, it does:
```
nqp::handle(
  nqp::call(blorst),
  'WARN',
  nqp::resume(nqp::exception())
)
```
So it's completely compile time and not amendable.  Which I find to be a disadvantage, as you might want to have warnings (not) enables in production vs development.

**Comment by vrurg** [comment]
> But the catching of a warning in practice is pretty complicated, needing a `CONTROL` block.

I don't see how is it complicated. Two more lines if `CATCH` and `CONTROL` are to be used within same block?

> So it's completely compile time and not amendable.

Only from the outside of it, which is totally correct. Consider the following:

```raku
CONTROL {
    when CX::Warn {
        note "OUTSIDE: ", .raku;
        .rethrow;
    }
}

sub foo {
    CONTROL {
        when CX::Warn {
            note "INSIDE: ", .raku;
            .rethrow;
        }
    }
    warn "testing warnings";
}

quietly foo;
```

It does what I would expect from the code: reports `INSIDE`, but not `OUTSIDE`.

Basically, speaking of the design side, with proper separation of concerns the code which throws warnings, and the code which handles them must not depend on the kind of warning to be thrown. With the only exception, perhaps, that the throwing part might chose the particular class to be thrown, depending on context. And then we should let the warning class to determine the kind of behavior it implements. This would allow users to have their own approach if necessary without requiring any changes to the core.

The optimizer might turn the course of events inside out if it knows that external locus of control would do better in a particular situation. But this is what optimziers do – they break rules intentionally in a controllable way. :)

**Comment by lizmat** [comment]
Compare:
```
sub foo {
    my $*WARNINGS = CX::Warn::Fatal;
    warn "testing warnings";
}
```
with:
```
sub foo {
    CONTROL {
        .throw;
    }
    warn "testing warnings";
}
```
Also: consider the possibility to switch on / off behaviour of warnings in a production environment.  How would you do that with CONTROL blocks?

**Comment by lizmat** [comment]
Come to think of it, we could also accept a `Block` as a valid value in `$*WARNINGS`:
```
sub foo {
    my $*WARNINGS = { .say }
    warn "foo";
}
```

**Comment by vrurg** [comment]
What advantage would it give to us? I'd stick to the point that a warning is a control exception and nothing else.

**Comment by niner** [comment]
> Also: consider the possibility to switch on / off behaviour of warnings in a
> production environment.  How would you do that with CONTROL blocks?

Isn't that as simple as `CONTROL { when CX::Warn { .gist.note unless 
%*ENV<PRODUCTION>; .resume } }`?

**Comment by vrurg** [comment]
> Isn't that as simple as `CONTROL { when CX::Warn { .gist.note unless %*ENV<PRODUCTION>; .resume } }`?

What is good about this PR is the ability to suppress noise when there is an external tool which produces harmless warnings but these warnings disturb me or my tool set in one or other way.

**Comment by lizmat** [comment]
I will rework this to use the CONTROL mechanism to keep compatibility

File: src/core.c/control.pm6 [hunk 2]
```diff
@@ -261,13 +262,12 @@ multi sub die(|cap ( *@msg ) --> Nil) {
...
 
 proto sub warn(|) {*}
-multi sub warn(*@msg) {
-    my $msg := @msg.join || "Warning: something's wrong";
-    my $ex := nqp::newexception();
-    nqp::setmessage($ex, nqp::unbox_s($msg));
-    nqp::setextype($ex, nqp::const::CONTROL_WARN);
+multi sub warn(*@msg --> 0) {
+    my str $message = @msg.join || "Warning: something's wrong";
+    my $ex := nqp::newexception();
```
Comment (MasterDuke17) [inline src/core.c/control.pm6:267]: Looks like some trailing whitespace got added here.

**Comment by vrurg** [comment]
Something went very much wrong about this PR's branch.

**Comment by lizmat** [comment]
I thought I had just rebased it  :-(

**Comment by lizmat** [comment]
I've recreated the branch and added a `CX::Warn::Debug` class.

**Comment by vrurg** [comment]
I still disagree with the course taken. To sum up, what worries me are:

1. Namespace inconsistency. `CX` stands for Control eXception, but the new classes are not `X::Control`. They are not even exceptions.
2. Complications for user code to handle warnings depending on what `$*WARNINGS` is set to. Instead of simply dispatching over exception type a handler would have to analyze on only the exception itself but `$*WARNINGS` itself.
3. I don't really see a use in supporting callables except for extra code to handle them. Same functionality could be achieved by a custom class, inheriting from CX::Warn.

I stick to the point that `$*WARNINGS` must be processed by `sub EXCEPTION`. Each `CX::Warn::` must be a child of `CX::Warn` and be (re-)throwabable by user code. `print_control` would only invoke `.WARN` and doesn't care about what it does but keep in mind that the method may not return.

```raku
CONTROL {
    when CX::Warn::Debug { ... }
    when CX::Warn::Fatal { ...; .rethrow }
}
```

```raku
class MyApp::CX::Warn::Logging is CX::Warn {
    method WARN {
        $logger.log: $.message;
        nextsame; # Let the default warning do its job
    }
}
```

**Review by vrurg — CHANGES_REQUESTED** [review]

File: src/core.c/Exception.pm6 [hunk 3]
```diff
@@ -428,28 +428,68 @@ my class CX::Last does X::Control {
...
+    method WARN(str $message, $bt --> Nil) is hidden-from-backtrace {
+        note "$message\n$bt.first-none-setting-line()";
+    }
 }
-my class CX::Succeed does X::Control {
-    method message() { "<succeed control exception>" }
-}
-my class CX::Proceed does X::Control {
-    method message() { "<proceed control exception>" }
+my class CX::Warn::Quietly {
```
[the remark this answers is not in the document]
Comment (lizmat) [reply src/core.c/Exception.pm6:442]: How does `&EXCEPTION` even enter into this?

The `CX::Warn::xxx` classes are not to be thrown.  The reason I chose their name that way, was a. to prevent more global name pollution.  I guess they could well be called `Warnings::Quietly`, `Warnings::Fatal`, etc.  Also, there's a bit of prior art with `$*UPGRADE-RAT` which currently takes the `CX::Warn` class to warn.

To re-iterate: the only time when `$*WARNINGS` is checked, is when a warning has already been thrown, and we're about to print the warning on STDERR.  Only **THEN** does it act differently (potentially) from before.

If your main beef is with the name of the `CX::Warn::xxx`, I can change them

[the remark this answers is not in the document]
Comment (vrurg) [reply src/core.c/Exception.pm6:442]: > To re-iterate: the only time when `$*WARNINGS` is checked, is when a warning has already been thrown, and we're about to print the warning on STDERR. Only **THEN** does it act differently (potentially) from before.

This is the key point I consider wrong.

Once again, the idea is to let user code decide their behavior depending on the current kind of warning. Best this is done if they can multi-dispatch over warning type object. To have this working correctly, as I stated, all of them must inherit from `CX::Warn` and conversion from VM-level control exception into a HLL `CX::Warn` instance must take place in `EXCEPTION`. In this case all `print_control` would have to do is invoke `WARN` method on a received `CX::Warn` instance. And when done this way the code in my examples would become possible.

[the remark this answers is not in the document]
Comment (lizmat) [reply src/core.c/Exception.pm6:442]: > conversion from VM-level control exception into a HLL CX::Warn instance must take place in EXCEPTION

It has never done so in the past.  I fail to see why now.

As to:
````raku
CONTROL {
    when CX::Warn::Debug { ... }
    when CX::Warn::Fatal { ...; .rethrow }
}
````
the whole point to *me* was not needing clunky `CONTROL` blocks.  And how would you throw a `CX::Warn::Fatal` in the first place??  What am I missing?

The whole idea was to get some sort of *global* control on the behaviour of `warn` and `CX::Warn` control exceptions thrown in another way.  Similar to how `%*ENV<RAKU_EXCEPTIONS_HANDLER>` allows you to specify how fatal exceptions are to be handled.  But with a lexical dynamic override.  And the possibility of `CONTROL` to still take local control about what to do any given warnings.

The example of the `CONTROL` block doesn't make any sense anyway, as in a given scope, you would only have one behaviour of warnings active anyway (unless you start doing *really* weird things as setting the value of `$*WARNINGS` multiple times in the same scope.

So in short: I'm not getting what your examples would actually add in useful functionality, other than to make them possible?

[the remark this answers is not in the document]
Comment (vrurg) [reply src/core.c/Exception.pm6:442]: So much to explain, so much has been understood incorrectly! :) 

> > conversion from VM-level control exception into a HLL CX::Warn instance must take place in EXCEPTION
> 
> It has never done so in the past. I fail to see why now.

[Current implementation]{https://github.com/rakudo/rakudo/blob/main/src/core.c/Exception.pm6#L476} does it for `CONTROL` blocks to work correctly. It would need to be slightly extended to support `$*WARNINGS` and that's all.

> the whole point to _me_ was not needing clunky `CONTROL` blocks.

The `CONTROL` example is not about throwing, it is about **handling** thrown exceptions. Once again, step back a little from your producing-centric point of view and try yourself in the shoes of one handling the warnings and wanting to have different reactions to different ones. What would they need to do in your current design? The `CONTROL` example is what they'd need in my case. Notice, please, that `$*WARNINGS` can be overridden (you stated it already) and same application can throw different kinds of warnings. Not necessarily that a warning thrown would be processed right away in the same `CONTROL` block, where it was caught, it could be postponed passed into a different thread or otherwise lose the context – hence, lose access to the original `$*WARNING`.

My approach doesn't require you to use `CONTROL` to throw/re-throw warnings, it preserves the whole thing about having `$*WARNING`. It just lets one to have more elegant, logical, and concise handling code.

> The example of the `CONTROL` block doesn't make any sense anyway, as in a given scope, you would only have one behaviour of warnings active anyway (unless you start doing _really_ weird things as setting the value of `$*WARNINGS` multiple times in the same scope.

Wrong because `CONTROL` may sit very early in the call stack. `$*WARNING` can be changed deep down the call chain, even by a 3rd-party module.

> And how would you throw a `CX::Warn::Fatal` in the first place??

```raku
class W::Foo is CX::Warn {
    method message { "Foo!" }
}

{
    CONTROL {
        note "OK? ", .raku;
        when W::Foo {
            note "It's W::Foo";
            .resume
        }
    }

    my $a = 42;
    W::Foo.new.throw if $a > 30;
    $a = 5;
    say $a;
}
```

This is something I played with a while ago. The only problem this code currently has is that the core doesn't know how to handle `W::Foo` because no one considered a possibility of inheriting from `CX::Warn`. So, replacing `.resume` with `.rethrow` or simply removing the `CONTROL` block results in "control exception without handler". I just haven't got time to take care of it.

> So in short: I'm not getting what your examples would actually add in useful functionality, other than to make them possible?

I hope I stated it above clearly, but worth repeating: simplify **handling** of different kinds of warnings thrown would it be required. Even without considering manipulations with `$*WARNINGS', I'd feel much better having to multi-dispatch over an exception type rather than analyzing what's in a dynamic variable which value isn't even guaranteed to be preserved for the time, when I start processing it.

[the remark this answers is not in the document]
Comment (lizmat) [reply src/core.c/Exception.pm6:442]: But, a `warn "foo"` doesn't call `EXCEPTION` at all????  Only if you have a `CONTROL { }` block active, does it run `EXCEPTION`?

[the remark this answers is not in the document]
Comment (lizmat) [reply src/core.c/Exception.pm6:442]: Not sure what you meant there, but that's the reason for *me* to put the check on `$*WARNINGS` in `print_control`.

Now, if you find the `CX::Warn` class naming confusing, I can change that.

[the remark this answers is not in the document]
Comment (vrurg) [reply src/core.c/Exception.pm6:442]: No, it's not about the naming. It is about **handling** by user code would that be necessary. It is about being able to multi-dispatch within `CONTROL` on a caught warning exception.

Simple, in your implementation `print_control` reads `$*WARNINGS` and interprets it. `EXCEPTION` behavior is unchanged, it creates a `CX::Warn` for user when needed.

I propose to teach `EXCEPTION` to read and interpret `$*WARNINGS`. `print_control` might use this ability to get a `CX::Warn` to call `WARN` upon. This way `$*WARNINGS` interpretation is localized in one place, but the result of it can be used by either user's `CONTROL` or by `print_control` equally well.

[the remark this answers is not in the document]
Comment (vrurg) [reply src/core.c/Exception.pm6:442]: And, BTW, I don't want them to be renamed. :)
